pkg/registry/servicecatalog/instance/strategy.go

```diff
@@ -156,7 +156,8 @@ func (instanceRESTStrategy) PrepareForUpdate(ctx genericapirequest.Context, new,
 	newServiceInstance.Spec.ClusterServicePlanRef = oldServiceInstance.Spec.ClusterServicePlanRef
 
 	// Clear out the ClusterServicePlanRef so that it is resolved during reconciliation
-	if newServiceInstance.Spec.ClusterServicePlanExternalName != oldServiceInstance.Spec.ClusterServicePlanExternalName {
+	if newServiceInstance.Spec.ClusterServicePlanExternalName != oldServiceInstance.Spec.ClusterServicePlanExternalName ||
+		newServiceInstance.Spec.ClusterServicePlanName != oldServiceInstance.Spec.ClusterServicePlanName {
```

Comment:
It feels like we should ignore the ExternalPlanName at first thought. Why do we have to care they match?

Reply:
The user can specify the class and plan using *either* the external names or the k8s names.

If the plan names do not match, that indicates that the user has requested a change to the plan. In that case, the plan that the ServiceInstance referenced is no longer the plan that the ServiceInstance should reference. The API Server cannot resolve the reference, so the API Server clears out the reference. The Controller then attempts to resolve the reference during reconciliation of the spec changes made to the ServiceInstance.